### ISSUE TITLE
Live code analyzer to highlight dangerous usage

### ISSUE THREAD

**Issue by medmondson** [FakeItEasy/FakeItEasy#723]
As we know faking an object containing non-virtual members will cause the actual code contained within those members to run.

With the advent of [Roslyn](https://github.com/dotnet/roslyn) it's now possible to provide 'live' code analysis as we type.  We could use this to warn a user of a potentially dangerous fake.

This would hold huge advantage to both new and advanced users.

I have experience in this area, therefore if you ratify my suggestion I'll be happy to begin working on this.

**Comment by thomaslevesque** [comment]
Hi @medmondson,

Thanks for the suggestion. It's an interesting idea; the only thing that bothers me is that there are perfectly valid cases where you would want to fake a class that has non-virtual members, so it would cause a lot of false positives.

BTW, we already have an analyzer package, which will be released very soon along with FakeItEasy 2.0.0. Currently it only checks that call specifications are configured or asserted (e.g. a statement like `A.CallTo(() => foo.Bar());` is obviously an mistake, since it has no useful effect), but we could add other diagnostics.

**Comment by adamralph** [comment]
I agree with @thomaslevesque that the problem would be false positives. Do Roslyn analyzers have the concept of "levels" of warnings/errors. This could only ever be a very light warning, perhaps just a "hint".

One thing we could do is catch the case where someone tries to configure a non-virtual member, e.g.

```C#
var foo = A.Fake<IFoo>();
A.CallTo(() => foo.NonVirtualMethod()).Returns(42);
```

We know that will always fail.

**Comment by blairconrad** [comment]
I'm not against catching the case where someone tries to configure a non-virtual member, but (I think) we already catch that at runtime. Sure, compile time would be a nice to have, but I wonder whether it's worthwhile duplicating all the logics.

**Comment by medmondson** [comment]
All valid points, thanks for the feedback. The analyzer would be set to
'warning' level only (which appears as a squiggly green underline and
description on hover).  In addition it would only appear when the object
contains a non-virtual member that contains code.

The analyzer would never fail the build and only appear when actual code
will be run on the object. This should mean no false positives.

Hope this helps?

**Comment by blairconrad** [comment]
I'm still not entirely sold on people perhaps legitimately faking a class that has non-virtual members getting squiggles all over. Perhaps I still don't understand fully.

> The analyzer would never fail the build and only appear when actual code
will be run on the object. This should mean no false positives.

Sorry, @medmondson, I don't understand what "only appear when actual code will be run on the object" means. Could you elaborate?

**Comment by thomaslevesque** [comment]
> Do Roslyn analyzers have the concept of "levels" of warnings/errors?

Yes. We could set the level to `Info`; the user can always override it in the .ruleset file if necessary.

> One thing we could do is catch the case where someone tries to configure a non-virtual member, e.g.
> 
>     A.CallTo(() => foo.NonVirtualMethod()).Returns(42);

But we already throw in that case, so an analyzer for this wouldn't be very useful...


> The analyzer would be set to 'warning' level only (which appears as a squiggly green underline and description on hover).

Many people compile with the "warnings as errors" option; if you have warnings, you should either fix or suppress them. So the false positives remain an issue.

> Sorry, @medmondson, I don't understand what "only appear when actual code will be run on the object" means. Could you elaborate?

Same here, I don't understand what you mean.

**Comment by medmondson** [comment]
No problem, totally understandable.

The underline would only appear where the fake is instantiated (so not
everywhere literally just the 'Candy' in A.Fake<Candy>();)

When I'm referring to actual code being ran I mean any code contained
within 'Candy' that is within a non-virtual member.

As this is impossible to fake even though the user believes they've created
a fake the actual implementation of 'Candy' would be executed - which is
both unintended and dangerous.

Imagine this like a bespoke resharper rule that ships within the project
with no external dependencies.

I'm still not entirely sold on people perhaps legitimately faking a class
that has non-virtual members getting squiggles all over. Perhaps I still
don't understand fully.

The analyzer would never fail the build and only appear when actual code
will be run on the object. This should mean no false positives.

Sorry, @medmondson <https://github.com/medmondson>, I don't understand what
"only appear when actual code will be run on the object" means. Could you
elaborate?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/FakeItEasy/FakeItEasy/issues/723#issuecomment-221347041>

**Comment by adamralph** [comment]
@medmondson the problem is, there are types with both virtual and non-virtual members. It's perfectly valid to fake on of those. Perhaps you mean a type which has *only* non-virtual members? Perhaps that  could work, although we'd still get a runtime error later whenever a configuration or assertion is attempted.

**Comment by blairconrad** [comment]
@medmondson, thanks for the response, although I think some weird formatting thing happened, as it seems to have actual replies and quotes from previous comments all jumbled up in a weird order. I think I understand much of your response, but am still struggling with

> When I'm referring to actual code being ran I mean any code contained
within 'Candy' that is within a non-virtual member.

So the Roslyn analyzer would have to know when the production code ends up calling a non-virtual member of the fake `Candy` object, and only then would it highlight the term `Candy` in the fake creation? This sounds very difficult, but would make the diagnostic more appealing to me.

Also, you said

> even though the user believes they've created
a fake the actual implementation of 'Candy' would be executed - which is
both unintended and dangerous.

It can be tricky, but I'm not sure it's always dangerous or unintended. While it's my preference to fake interfaces only, as @adamralph says, if a class has both virtual and non-virtual members, there could be legitimate reasons for faking it.

**Comment by medmondson** [comment]
Yes, perhaps that would make more sense being and object with entirely non-virtuals. Agree with what you say but surely it's better to flag a potential problem when coding rather than runtime.

**Comment by blairconrad** [comment]
If the code were nearly always prone to causing a problem, I'd be in agreement. I just worry about the rate of false positives.

**Comment by medmondson** [comment]
In light of all comments I'll look at putting some examples together covering each scenario raised and my expectations in each scnerio. Feel free to keep adding concerns it's great to have such an active and passionate community!

**Comment by blairconrad** [comment]
Thanks, @medmondson. Examples may be helpful.

**Comment by blairconrad** [comment]
>  I'll look at putting some examples together covering each scenario raised and my expectations in each scnerio. 

Hey, @medmondson, have you come up with anything? So far I don't think we have anything actionable. If there's nothing else to add, I'm inclined to close this issue. It can always be reopened in the future. Or other, more specific issues could be raised, covering particular dangerous usages.

**Comment by blairconrad** [comment]
@medmondson, no need to apologise, and I didn't mean to pressure you. I was just doing some issue housecleaning If you're still interested, I'm happy to keep the issue open. Good luck with the move!

**Comment by medmondson** [comment]
Okay - I know it's been a while.  Thanks for bearing with me.

I've put together an implementation of an analyzer along with example useage of FIE.  I've named them "AnalyzeItEasy" and "AIESubject" respectively (no points for the names!) and [both are available here](https://github.com/medmondson/AnalyzeItEasy).

I've tried to take all concerns raised above into consideration, that being:

- We want to warn the user if they fake a class composed of entirely non-virtual members.
- There should be no false positives.
- It should be a warning. Which by default will not fail the build.

To do this I created an example of each and faked it.

- IMessageSender, an interface
- VirtualMessageSender, a class made of entirely **virtual** members
- MixedMessageSender, a class made of both **virtual and non-virtual** members
- NonVirtualMessageSender, a class made of entirely **non-virtual** members

![examplecodeanalysis](https://cloud.githubusercontent.com/assets/1955888/16902908/92d5f2f6-4c65-11e6-9ed0-c1f753e74312.png)

Hopefully you can see the only warning is produced on the entirely non-virtual member class. Meaning that _any fake_ produced of of this class will mean the actual implementation of that class will be executed.  Therefore a warning is appropriate.

Please take the time to [understand the code](https://github.com/medmondson/AnalyzeItEasy) and feel free to comment on here or submit PRs for anything I might have overlooked.

Once again thanks for waiting - I literally just got the internet connected at my new house :-)

**Comment by blairconrad** [comment]
Thanks for the update, @medmondson.

I hate to be a party-pooper, but to start, I don't think there's such a thing as a non-virtual member class `NonVirtualMessageSender` has 3 virtual members: `GetHashCode`, `Equals`, and `ToString`. So it seems that everything's a false positive, in a sense. You may argue that those 3 method don't count, but some users may think they do. Worse, if you create a class like this: 

```c#
public class DerivedVirtualMessageSender:VirtualMessageSender
{
}
```
and try to fake it, it shows the warning, even though the class has member `virtual Send(MailAddress recipient, string messageBody)`.

Ultimately, I fear that analyzing only members _defined_ on a class will end up missing base class members, and so give false positives.
And following the inheritance chain will always reveal a virtual member. Accordingly, I'm not sure the analyzer can provide value.

**Comment by medmondson** [comment]
As you suspected I'd never considered those base methods. However I am certain there is at least _some_ value here.  Remember the user can always override it in the .ruleset file if necessary.

Perhaps my analyzer is too ambiguous and should instead concentrate on the specific set-up calls of a non-virtual member. Let's assume `GetTopSellingCandy` is non-virtual:

`A.CallTo(() => shop.GetTopSellingCandy()).Returns(lollipop); //Warning here`

It's my understanding this would never be a valid mock yet would compile and silently execute the implementation, which brings us right around to the comment from @thomaslevesque.

Although I understand we would throw here the trend has always being in bringing potential problems closer to the developer, meaning we're causing the developer to have to run the code to find the problem, even though it could never have been valid.

Really appreciate the time you guys are spending on this. Any thoughts?

**Comment by adamralph** [comment]
> Perhaps my analyzer is too ambiguous and should instead concentrate on the specific set-up calls of a non-virtual member.

I think that could work.

**Comment by thomaslevesque** [comment]
Thanks @medmondson for taking the time to make this POC.

> Perhaps my analyzer is too ambiguous and should instead concentrate on the specific set-up calls of a non-virtual member. Let's assume `GetTopSellingCandy` is non-virtual:
> 
>     A.CallTo(() => shop.GetTopSellingCandy()).Returns(lollipop); //Warning here
>
> It's my understanding this would never be a valid mock yet would compile and silently, which brings us right around to the comment from @thomaslevesque.

Indeed. This code would compile, but would throw at runtime.

> Although I understand we would throw here the trend has always being in bringing potential problems closer to the developer, meaning we're causing the developer to have to run the code to find the problem, even though it could never have been valid.

I would agree if it were production code; but in unit tests, it's different, because you typically run them immediately after writing them, and the exception message states clearly enough what the problem is.

Also, unless I'm missing something, your analyzer doesn't check which method is being called in the `InvocationExpression`; this means the warning will also trigger for calls that have nothing to do with FakeItEasy, e.g. `MyGenericMethod<NonVirtualMessageSender>()`. It should be limited to `A.Fake<T>()` (of course, this could easily be fixed).

Overall, I'm still not convinced this analyzer is a good idea. As mentioned by @blairconrad, *all* types have virtual members, unless `Equals`/`GetHashCode`/`ToString` are specifically overriden to make them `sealed`. Even if you explicitly exclude these methods, the warning will be triggered only for classes that have only non-virtual members, which has little value IMO. If you have 99 non-virtual methods and 1 virtual method, the warning won't appear, but it will still fail when you try to configure a non-virtual one.

Now, if I understand correctly, your main concern is that the code using the fake might end up calling a non-configured and non-virtual method, which would result in calling the actual implementation. But this can happen even with classes that have both virtual and non-virtual members, so your analyzer wouldn't catch this case. Even if you try to analyze which members are configured, the issue still can't be caught, since it's not possible to configure non-virtual members. Just because a member isn't configured doesn't mean it isn't called... We also can't make fakes throw when a non-virtual member is called, because we can't intercept the call.

One thing we *could* do is this: when a fake is created with the `Strict` option, throw if the type has non-virtual members, because we can't guarantee that there won't be calls to unconfigured members. Oh, and this would be a breaking change (but one that would probably make sense; currently, strict fakes are not as strict as they seem, because of non-virtual members).

**Comment by blairconrad** [comment]
Indeed, thanks for the proof of concept, @medmondson.

 > Perhaps my analyzer is too ambiguous and should instead concentrate on the specific set-up calls of a non-virtual member.
> …
> Although I understand we would throw here the trend has always being in bringing potential problems closer to the developer, meaning we're causing the developer to have to run the code to find the problem, even though it could never have been valid.

I also think there's some value in this approach. Not a tremendous amount, since as @thomaslevesque points out, there a (fairly clear, I think) exception that's thrown when the call is configured, and in theory the exception should be raised second (or maybe a few minute) after the code has been written, so I don't expect users to be too confused.
That being said, I enjoy my red "this won't compile" squiggles in the IDE before I try to compile, so I see the attraction. I don't see enough value for it to pique my interest enough to try to write an analyzer myself, but if a working analyzer fell into my lap, I wouldn't turn it away.

As @thomaslevesque points out, the analyzer currently doesn't limit itself to `A.Fake` calls, but I assumed that you'd just not yet implemented that parts as part of the prototype, and that it would be added later. Just in case, it's also worth mentioning that the warning message complained about the type `A`, not `NonVirtualMessageSender`.

As for @thomaslevesque's comment

> One thing we could do is this: when a fake is created with the Strict option, throw if the type has non-virtual members, because we can't guarantee that there won't be calls to unconfigured members.

This is a good point, and one that I never really thought of. I'm not eager to make the change, though, as it would keep the advanced users from legitimately faking a mixed-member class when they knew what they were doing and didn't care about the non-virtual members. Also, keep in mind that [GetType](https://msdn.microsoft.com/en-us/library/system.object.gettype(v=vs.110).aspx) is not virtual. :wink:

**Comment by thomaslevesque** [comment]
> This is a good point, and one that I never really thought of. I'm not eager to make the change, though, as it would keep the advanced users from legitimately faking a mixed-member class when they knew what they were doing and didn't care about the non-virtual members.

True... maybe introduce a `VeryStrict` option? :stuck_out_tongue:


> Also, keep in mind that GetType is not virtual.

And `MemberwiseClone` too

**Comment by blairconrad** [comment]
> True... maybe introduce a `VeryStrict` option? :stuck_out_tongue:

I'd rather just turn off faking for anything but interfaces. It'd solve _a lot_ of problems. Except for the problem of not being able to fake anything but an interface. I can dream, right?

> And `MemberwiseClone` too

touché

**Comment by medmondson** [comment]
> I don't see enough value for it to pique my interest enough to try to write an analyzer myself, but if a working analyzer fell into my lap, I wouldn't turn it away.

Assuming this is the consensus I'd be more than happy to work towards this

**Comment by thomaslevesque** [comment]
> Assuming this is the consensus I'd be more than happy to work towards this

Just to make sure we're talking about the same thing: you mean detecting attempts to configure/assert non-virtual members? As I said, I'm not really convinced of the usefulness (since it's going to throw anyway), but I see no harm in it either. You can probably reuse parts of the [existing analyzer](https://github.com/FakeItEasy/FakeItEasy/blob/master/src/FakeItEasy.Analyzer/UnusedReturnValueAnalyzer.cs) to identify the method being called.

**Comment by blairconrad** [comment]
And even more to make sure we're talking about the same thing, @medmondson, you're planning on submitting a PR for FakeItEasy.Analyzer, yes? Not making up a whole new analyzer with a vsix, like you did for your sample?

**Comment by medmondson** [comment]
@thomaslevesque - yes that's what I'll work towards, we'll forget my ambiguous class-based POC

@blairconrad - correct, I'll update the analyzer that already exists (I didn't realize one existed when I started)

**Comment by blairconrad** [comment]
The fix for this issue has been merged as #848 (which superseded #826).

@medmondson, thanks for all your hard work on this. We appreciate it!

**Comment by medmondson** [comment]
Thanks @blairconrad and @thomaslevesque. I really enjoyed working on this and glad I could help. If there is anything else at all wether it be big/small/tedious/anything please let me know and feel free to assign it to me. I'm more than happy to help.

Thanks again for taking my idea and code on board. Looking forward to further contributions.

**Comment by blairconrad** [comment]
@medmondson, we enjoyed working with you as well. Thanks very much for the offer of help. Accepted!
But if I may suggest, it's probably better for you to "pull" issues, either from those already in the list, or some itch you want scratched. I wouldn't feel right just assigning arbitrary work to you that may not align with your current availability or interests. So if you see something, just hoot us, and it's yours.

**Comment by blairconrad** [comment]
This issue has been released in [FakeItEasy 2.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.3.0).

@medmondson, thanks again, and look for your name in the release notes! :trophy:

**Comment by medmondson** [comment]
Great, glad to see it! Already looking forward to contributing more in the future :-)
